website/docs/internals/json-format.html.md

```diff
@@ -0,0 +1,481 @@
+---
+layout: "docs"
+page_title: "Internals: JSON Output Format"
+sidebar_current: "docs-internals-json"
+description: |-
+  Terraform provides a machine-readable JSON representation of state, configuration and plan.
+---
+
+# JSON Output Format
+
+-> **Note:** This format is available in Terraform 0.12 and later.
+
+When Terraform plans changes, it prints a human-readable summary to the terminal. It can also, when run with `-out=<PATH>`, write a much more detailed binary plan file, which can later be used to apply those changes.
+
+Since the format of plan files isn't suited for use with external tools (and likely never will be), Terraform can output a machine-readable JSON representation of a plan file's changes. It can also convert state files to the same format, to simplify data loading and provide better long-term compatibility.
+
+Use `terraform show -json <FILE>` to generate a JSON representation of a plan or state file. See [the `terraform show` documentation](/docs/commands/show.html) for more details.
+
+-> **Note:** The JSON output format is experimental and subject to change. A `"format_version"` key is included in the output. The minor version portion will be incremented for compatible additions to the output, and the major version portion will be incremented for any format changes which require changes for correct processing.
```

Comment:
Oh, I didn't notice this yesterday either, but: where actually IS the `"format_version"` key? It's not in the example output yet. I'd assume it's in the top-level state or plan reps?

Reply:
Another detail we may wish to mention here is that we've initially numbered it `0.1` with the intent of saying "this is experimental and there are no compatibility guarantees"; the rules about incrementing the major/minor versions here will apply only after we number it `1.0`, which we'll do after this format has proven itself with a number of different use-cases.

That's a lot to write in a note, so perhaps for now we could just say something like "The `format_version` currently has major version zero to indicate that the format is experimental and subject to change. A future version will assign a non-zero major version and make stronger promises about compatibility. We do not anticipate any significant breaking changes to the format before its first major version, however."

(In other words, can we wait to document the specific major/minor distinction until we get to format 1.0, since it's a moot point until then anyway?)